src/app/shared/services/upload/upload.service.ts

```diff
@@ -0,0 +1,143 @@
+import { HttpErrorResponse } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import {
+  BehaviorSubject,
+  combineLatest,
+  EMPTY,
+  from,
+  of,
+  throwError,
+  timer,
+} from 'rxjs';
+import {
+  catchError,
+  concatMap,
+  debounceTime,
+  distinctUntilChanged,
+  filter,
+  map,
+  mergeMap,
+  retryWhen,
+  switchMap,
+  tap,
+} from 'rxjs/operators';
+import { DiaBackendAssetRepository } from '../dia-backend/asset/dia-backend-asset-repository.service';
+import { getOldProof } from '../repositories/proof/old-proof-adapter';
+import { Proof } from '../repositories/proof/proof';
+import { ProofRepository } from '../repositories/proof/proof-repository.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class UploadService {
+  private readonly _isPaused$ = new BehaviorSubject(false);
+  private readonly _isPausedByFailure$ = new BehaviorSubject(false);
+  private readonly _taskQueue$ = new BehaviorSubject<Proof[]>([]);
+  private readonly _currentUploadingCount$ = new BehaviorSubject(0);
+  readonly isPaused$ = this._isPaused$
+    .asObservable()
+    .pipe(distinctUntilChanged());
+  readonly isPausedByFailure$ = this._isPausedByFailure$
+    .asObservable()
+    .pipe(distinctUntilChanged());
+  readonly currentUploadingCount$ = this._currentUploadingCount$
+    .asObservable()
+    .pipe(distinctUntilChanged());
+  private readonly taskQueue$ = combineLatest([
+    this._taskQueue$.asObservable().pipe(distinctUntilChanged()),
+    this.isPaused$,
+  ]).pipe(map(([t, _]) => t));
+
+  constructor(
+    private readonly diaBackendAssetRepository: DiaBackendAssetRepository,
+    private readonly proofRepository: ProofRepository
+  ) {}
+
+  initialize$() {
+    return combineLatest([
+      this.uploadTaskDispatcher$(),
+      this.uploadTaskWorker$(),
+    ]);
+  }
+
+  pause() {
+    this._isPaused$.next(true);
+  }
+
+  resume() {
+    this._isPaused$.next(false);
+  }
+
+  private uploadTaskDispatcher$() {
+    const taskDebounceTime = 50;
+    return combineLatest([
+      this.proofRepository.getAll$().pipe(debounceTime(taskDebounceTime)),
+      this.isPaused$,
+    ]).pipe(
+      tap(([proofs, isPaused]) => {
+        const tasks = proofs.filter(
+          proof => !proof.diaBackendAssetId && !proof.willCollectTruth
+        );
+        this._currentUploadingCount$.next(tasks.length);
```

Comment:
I think we should update the `this._currentUploadingCount` in [`this.uploadTaskQueue`](https://github.com/numbersprotocol/capture-lite/pull/451/files#diff-566b2bc9075a195ca816285e30a40af05260703c2aaf60a93dc82a06fc67cbe9R107).

Also, why do we need `this._currentUploadingCount$`? I think generating `this.currentUploadingCount$` from `this.taskQueue$` could be a good idea?

Reply:
The initial design is different from the current implementation, so it's a good suggestion I overlooked.

Reply:
The `currentUploadingCount$` holds the task count that can be exposed and displayed on UI and it should not be cleared when pausing, while the taskQueue$ will be cleared when pausing. The component also uses `currentUploadingCount$` to determine if all tasks are completed, so filtering out zero value cannot be a solution.

I haven't come up with a decent solution yet, but I think this could be improved later.